client/components/FormattingBar/media/MediaTwitter.tsx

```diff
@@ -0,0 +1,119 @@
+import React, { Component } from 'react';
+import { InputGroup, Button, Intent, NonIdealState } from '@blueprintjs/core';
+import { isHttpsUri } from 'valid-url';
+
+import Icon from 'components/Icon/Icon';
+import { apiFetch } from 'client/utils/apiFetch';
+import { getIframeSrc, getEmbedType } from 'client/utils/editor';
+
+type Props = {
+	onInsert: (...args: any[]) => any;
+};
+
+const sampleUrl =
+	process.env.NODE_ENV === 'production'
+		? 'https://twitter.com/pubpub'
+		: 'https://twitter.com/kfutures';
+
+type State = any;
+
+class MediaTwitter extends Component<Props, State> {
+	constructor(props: Props) {
+		super(props);
+		this.state = {
+			isValid: false,
+			input: '',
+			embedUrl: '',
+			embedTitle: '',
+		};
+		this.handleInput = this.handleInput.bind(this);
+		this.handleInsert = this.handleInsert.bind(this);
+	}
+
+	handleInput(url) {
+		const input = url;
+		const isValid = isHttpsUri(input) && getEmbedType(input) === 'twitter';
+		this.setState(
+			{
+				input: input,
+				isValid: isValid,
+			},
+			() => {
+				if (!this.state.isValid) {
+					return this.setState({ embedUrl: '', embedTitle: '' });
+				}
+
+				const queryParams = `?type=${getEmbedType(input)}&input=${input}`;
+				// @ts-expect-error ts-migrate(2554) FIXME: Expected 2 arguments, but got 1.
+				return apiFetch(`/api/editor/embed${queryParams}`).then((result) => {
+					this.setState({
+						embedUrl: result.html,
+						embedTitle: result.title,
+					});
+				});
```

Comment:
I'm not totally sure that this need to be in `setState`'s callback function, since we already know the value of `isValid` before the component updates. This could be refactored into something like...

```typescript
const input = url;
const isValid = ...;

this.setState({ input: input, isValid: isValid });

if (!isValid) {
  // this and the above setState call will be batched into the same update
  this.setState({ embedUrl: '', embedTitle: '' });
  return;
}

const queryParams = ...

apiFetch(...).then(result => this.setState(...))

```

Reply:
good catch! We're gonna need to update this in all the embeds eventually.